### ISSUE TITLE
Extra array items when using data-unchecked-value

### ISSUE THREAD

**Issue by danchapmanme** [marioizquierdo/jquery.serializeJSON#67]
In this form example:

```
<input type="text"     name="question" value="Favorite color?">

<input type="checkbox" name="answers[][correct]:boolean" value="true" data-unchecked-value="false">
<input type="text"     name="answers[][text]" value="Blue">

<input type="checkbox" name="answers[][correct]:boolean" value="true" data-unchecked-value="false">
<input type="text"     name="answers[][text]" value="Green">

<input type="checkbox" name="answers[][correct]:boolean" value="true" data-unchecked-value="false">
<input type="text"     name="answers[][text]" value="Red">
```

If all checkboxes are ticked (checked), it returns the correct JSON:
```
{
  "question": "Favorite color?",
  "answers":[
    {"correct":true,"text":"Blue"},
    {"correct":true,"text":"Green"},
    {"correct":true,"text":"Red"}
  ]
}
```

But if none of the checkboxes are ticked, it returns invalid JSON:
```
{
  "question": "Favorite color?",
  "answers": [
    {"text":"Blue"},
    {"text":"Green"},
    {"text":"Red","correct":false},
    {"correct":false},
    {"correct":false}
  ]
}
```

**Comment by marioizquierdo** [comment]
This seems to be a legitimate problem with the `data-unchecked-value` option implementation with the "auto-arrays" syntax ('answers[]'). Probably because the option adds the values to the serialized JSON after all the values were parsed, therefore they are added in different order.

I'll take a look and see if it can be fixed without breaking other functionality. But for the time being, you can prevent this kind of behavior by making the names non deterministic, adding a explicit index on the "answers" array position:

```
<input type="checkbox" name="answers[0][correct]:boolean"  value="true" data-unchecked-value="false"
 />
<input type="text"     name="answers[0][text]" value="Blue" />

<input type="checkbox" name="answers[1][correct]:boolean"  value="true" data-unchecked-value="false"
 />
<input type="text"     name="answers[1][text]" value="Green" />

<input type="checkbox" name="answers[2][correct]:boolean"  value="true" data-unchecked-value="false"
  />
<input type="text"     name="answers[2][text]" value="Red" />
```

Then use the option `useIntKeysAsArrayIndex` to make sure the serialized object is an array:
```
$("input").serializeJSON({useIntKeysAsArrayIndex: true });

**Comment by marioizquierdo** [comment]
I will have to close this a WONTFIX.

This doesn't seem to be possible to fix without reimplementing `jQuery.serializeArray`, which is the foundation of this plugin.

The problem is that `jQuery.serializeArray` ignores unchecked checkboxes by design, which is the reason why the attr `data-uncheked-value` was added to this plugin in the first place.

This is how unckecked values is implemented (pseudocode):
```
formAsArray = jQuery(form).serializeArray(); // ignores unchecked checkboxes
for each unchecked-checkbox do
  formAsArray.push(serialized(unchecked-checkbox))
end
```

This has the problem of adding unchecked checkboxes at the end, which combined with the order-dependent array syntax (`name[]`) and nested objects (`name[][subname]`) adds the values in a different order. In other words, the inputs used in the example above are added like if they were declared in this order (if the checkboxes are unchecked):
```
// values added by jQuery.serializeArray
<input type="text"     name="answers[][text]" value="Blue" />
<input type="text"     name="answers[][text]" value="Green" />
<input type="text"     name="answers[][text]" value="Red" />

// unckeched checkboxes added later by serializeJSON
<input type="checkbox" name="answers[][correct]:boolean" value="true" data-unchecked-value="false"  />
<input type="checkbox" name="answers[][correct]:boolean" value="true" data-unchecked-value="false" />
<input type="checkbox" name="answers[][correct]:boolean" value="true" data-unchecked-value="false" />
```

And `serializeJSON` doesn't know where to place each key, because there's no reference to the order.

That's why the solution is to simply make the object positions explicit `answers[0], answers[1], answers[2]`, so the nested values are placed on the right position.


Here are some StackOverflow responses trying to deal with this problem in `serializeArray`: http://stackoverflow.com/questions/7335281/how-can-i-serializearray-for-unchecked-checkboxes. And some trying to deal with the same problem in Rails: http://stackoverflow.com/questions/20379311/how-to-get-blank-checkboxes-to-pass-as-false-to-params

So, this is a list of possible solutions:
  1. First, think if you really need those values. A missing key will return `undefined` and that's probably fine for your case.
  2. Add array indexes and use the option `useIntKeysAsArrayIndex: true` as suggested in the previous comment.
  3. Add hidden fields for each checkbox with the desired unchecked value (see https://github.com/marioizquierdo/jquery.serializeJSON/issues/19)
  4. You could also use jQuery to check all unchecked checkboxes before serialization, then uncheck them again (this makes an ugly flash in some browsers)

My preferred option is to use hidden fields, which is totally compatible with standard HTML.

**Comment by marioizquierdo** [comment]
This is fixed on #115 and will be released on version 3.2.0
